server/server/src/billing/billing_service.rs

```diff
@@ -0,0 +1,522 @@
+                        ClientError(SwitchAccountTierError::UserNotFound)
+                    }
+                    GetFileUsageError::Internal(e) => ServerError::from(e),
+                })?
+                .iter()
+                .map(|a| a.size_bytes)
+                .sum();
+
+            if usage > FREE_TIER_USAGE_SIZE {
+                return Err(ClientError(
+                    SwitchAccountTierError::CurrentUsageIsMoreThanNewTier,
+                ));
+            }
+
+            let (active_subscription, active_pos) =
+                get_active_subscription::<SwitchAccountTierError>(
+                    user_info.subscriptions.as_slice(),
+                )?;
+
+            stripe_client::delete_subscription(server_state, &active_subscription.id).await?;
+
+            user_info.subscriptions[active_pos].is_active = false;
+
+            FREE_TIER_USAGE_SIZE
+        }
+    };
+
+    let watched_keys = &[
+        stripe_user_info(&context.public_key),
+        data_cap(&context.public_key),
+    ];
+
+    let tx_result = tx!(&mut con, pipe_name, watched_keys, {
+        pipe_name
+            .set(data_cap(&context.public_key), new_data_cap)
+            .del(stripe_in_billing_workflow(&context.public_key))
+            .json_set(stripe_user_info(&context.public_key), &user_info)
+    });
+    return_if_error!(tx_result);
+
+    Ok(SwitchAccountTierResponse {})
+}
+
+fn get_active_subscription<E: std::fmt::Debug>(
+    subscriptions: &[StripeSubscriptionInfo],
+) -> Result<(StripeSubscriptionInfo, usize), ServerError<E>> {
+    let active_subscriptions: Vec<&StripeSubscriptionInfo> =
+        subscriptions.iter().filter(|info| info.is_active).collect();
+
+    if active_subscriptions.len() > 1 {
+        return Err(internal!(
+            "Redis says more than one stripe subscription is active: {:?}",
+            subscriptions
+        ));
+    }
+
+    let active_subscription = active_subscriptions
+        .get(0)
+        .ok_or(internal!("Redis says there is no active subscription despite the user having non free data cap: {:?}", subscriptions))?
+        .deref()
+        .deref()
+        .clone();
+
+    let active_pos = subscriptions
+        .iter()
+        .position(|info| info.is_active)
+        .ok_or(internal!("Redis says there is no active subscription despite the user having non free data cap: {:?}", subscriptions))?;
+
+    Ok((active_subscription, active_pos))
+}
+
+async fn lock_payment_workflow(
+    public_key: &PublicKey,
+    con: &mut deadpool_redis::Connection,
+) -> Result<(), ServerError<SwitchAccountTierError>> {
+    let tx_result = tx!(con, pipe_name, &[stripe_in_billing_workflow(public_key)], {
+        if con.exists(stripe_in_billing_workflow(public_key)).await? {
+            return Err(Abort(ClientError(
+                SwitchAccountTierError::CurrentlyInBillingWorkflow,
+            )));
+        }
+
+        pipe_name
+            .set(stripe_in_billing_workflow(public_key), 1)
```

Comment:
there is a `set_ex` command that will do this in one go, not a huge deal because it's a pipeline.
I think 5 seconds is too little, I think you should do like 30 seconds. I'm still not clear why there are 2 keys for this, but I will continue reading through to see.